### PR TITLE
Use external project all

### PR DESCRIPTION
**Pull request by fmessmer**
includes #81 
uses the ExternalProject concept for the other packages in this repo

### CODE REVIEW

**Comment by fmessmer** [comment]
@ipa-mdl 
I started to apply your concept to (for a start) `libphidgets`.
It seems to work, i.e. I can also successfully build `cob_phidgets` which depends on `libphidgets`...

Could you have a quick look at https://github.com/ipa320/cob_extern/pull/82/commits/170333fbb54b9bfad5a0f71b3a71cc1bd857cee8 and let me know whether this is the correct way to do it?

**Comment by mathias-luedtke** [comment]
You don't need to setup the soname chain, just copy the lib to the target.
The install script should get update, too.
(or just deleted)

**Comment by fmessmer** [comment]
> The install script should get update, too.
(or just deleted)

was wondering about this as well...the `install_pcan.sh` is not updated either...should I remove both?

**Comment by mathias-luedtke** [comment]
> was wondering about this as well...the install_pcan.sh is not updated either...should I remove both?

`install_pcan.sh` was not broken ;)

File: libntcan/CMakeLists.txt [hunk 1]
```diff
@@ -21,34 +21,7 @@ catkin_package(
...
-add_dependencies(ntcan build_ntcan)
-add_dependencies(ntcan build_ntcan3)
-add_dependencies(ntcan build_ntcan317)
-
-install(TARGETS ntcan
-    ARCHIVE DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
-    LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
-    RUNTIME DESTINATION ${CATKIN_GLOBAL_BIN_DESTINATION}
-)
+link_directories(common/lib/${MYARCH})
```
Comment (ipa-mdl) [inline libntcan/CMakeLists.txt]: this is not needed.

Comment (mathias-luedtke) [inline libntcan/CMakeLists.txt]: Better move this above line 18

File: libntcan/CMakeLists.txt [hunk 2]
```diff
@@ -21,34 +21,7 @@ catkin_package(
 )
 
 include_directories(common/include)
```
Comment (mathias-luedtke) [inline libntcan/CMakeLists.txt]: this is not needed

**Comment by fmessmer** [comment]
but still you could just use 
`cd $SOURCE_DIR && make install -C`
instead of
`cd $SOURCE_DIR && make install KERNEL_LOCATION=$KERNEL_VER DBG=NO_DEBUG MOD=MODVERSIONS PAR=PARPORT_SUBSYSTEM USB=USB_SUPPORT PCI=PCI_SUPPORT DNG=DONGLE_SUPPORT ISA=NO PCC=NO NET=NO RT=NO`

if I understood https://github.com/ipa320/cob_extern/commit/21bcb6847ed6b6601dbc9d709a2cb7548f0c7b7b correctly

**Comment by mathias-luedtke** [comment]
not sure why `libntcan` depends on `dpkg`

**Comment by fmessmer** [comment]
> not sure why `libntcan` depends on `dpkg`

for testing the architecture

**Comment by mathias-luedtke** [comment]
>> not sure why libntcan depends on dpkg

>for testing the architecture

Ah, we are calling dpkg directlly.
This does not work with other linux distros, we should consider moving to uname later.

**Comment by mathias-luedtke** [comment]
> but still you could just use
cd $SOURCE_DIR && make install -C
instead of
cd $SOURCE_DIR && make install KERNEL_LOCATION=$KERNEL_VER DBG=NO_DEBUG MOD=MODVERSIONS PAR=PARPORT_SUBSYSTEM USB=USB_SUPPORT PCI=PCI_SUPPORT DNG=DONGLE_SUPPORT ISA=NO PCC=NO NET=NO RT=NO

No, the make arguments are important.

**Comment by fmessmer** [comment]
I'm done for the time being...
I also tested this PR by having `cob_driver` and `autopnp` (uses `libconcorde_tsp_solver` and other packages) in an overlay using `cob_extern` from install space. `cob_driver` works fine...with `autopnp` I have some (non-cob_extern-related) issues as the workspace cannot be that minimal for compiling `autopnp`...maybe @ipa-rmb can help me test this in his setup tomorrow

@ipa-mdl do you have any additional comments?
#79 has been rebased in order to include #82...let's see whether Travis is happy

**Comment by fmessmer** [comment]
ToDo: convert architecture check from `dpkg` to `uname`

**Comment by fmessmer** [comment]
> ToDo: convert architecture check from `dpkg` to `uname`

@ipa-mdl
would `CMAKE_SYSTEM_PROCESSOR` be an option (https://cmake.org/cmake/help/v3.0/variable/CMAKE_SYSTEM_PROCESSOR.html)

**Comment by mathias-luedtke** [comment]
~~It's even better!
(and uses uname under the hood)~~

It is the processor architecture, but we need the machine architecture. (?)

**Comment by fmessmer** [comment]
So we would need to test this on various architectures...32bit, 64bit, linux, osx, arm...
what's the easiest way to test whether it works for all of them?

**Comment by mgruhler** [comment]
You need to test for 32bit and 64bit only, IMO. We are not supporting arm, let alone osx, afaik.

Thus, testint 32 vs 64 should be fine. This can easily be done in a VM.

**Comment by mgruhler** [comment]
@ipa-mdl so what uname command would you use?

`uname -p` `uname -i` and `uname -m` all return `x86_64`

But only `-m` seems to be support in a portable way:

```
       -m, --machine
              print the machine hardware name

       -p, --processor
              print the processor type (non-portable)

       -i, --hardware-platform
              print the hardware platform (non-portable)

```

**Comment by mathias-luedtke** [comment]
`uname -m` should be correct, but I don't know if this works correctly with docker / the buildfarm.

**Comment by mathias-luedtke** [comment]
`uname` does not work in docker. `dpkg` should work, but only for debian-based platforms.
Starting with kinetic the buildfarm build Fedora RPMs as well.

**Comment by fmessmer** [comment]
So I guess then it's easiest to keep the `dpkg`-struct as is...

**Comment by fmessmer** [comment]
@ipa-rmb 
Could you test this branch in your `autopnp` setup in order to guarantee that "your" packages still work as expected? Or let me know when we could do this together...

**Comment by ipa-rmb** [comment]
I will test it tomorrow.

**Comment by ipa-rmb** [comment]
libconcorde works fine with devel/setup.bash and install/setup.bash

**Comment by ipa-rmb** [comment]
lab_ipa_paths

**Pull request merged**